File: src/main/java/com/fendyk/clients/apis/LandAPI.java
package com.fendyk.clients.apis;

import com.fendyk.API;
import com.fendyk.DTOs.ChunkDTO;
import com.fendyk.DTOs.LandDTO;
import com.fendyk.DTOs.MinecraftUserDTO;
import com.fendyk.DTOs.TaggedLocationDTO;
import com.fendyk.Main;
import com.fendyk.clients.ClientAPI;
import com.fendyk.clients.fetch.FetchLand;
import com.fendyk.clients.redis.RedisLand;
import net.luckperms.api.model.user.User;
import org.bukkit.Bukkit;
import org.bukkit.Chunk;
import org.bukkit.Location;
import org.bukkit.entity.Player;
import org.joda.time.DateTime;

import java.util.Calendar;
import java.util.Date;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;

public class LandAPI extends ClientAPI<FetchLand, RedisLand, String, LandDTO> {

    public LandAPI(API api, FetchLand fetch, RedisLand redis) {
        super(api, fetch, redis);
    }

    /**
     * Creates a new land. Requires an existing chunk
     * @param owner
     * @param name
     * @param chunk
     * @return The new land if created or null if something went wrong.
     * @throws Exception When something goes wrong lol
     */
    public LandDTO create(Player owner, String name, Chunk chunk, Location location) throws Exception {
        UUID uuid = owner.getUniqueId();
        User user = Main.getInstance().getLuckPermsApi().getUserManager().getUser(uuid);

        if(user == null) return null;

        /* Get references */
        ChunkAPI chunkAPI = api.getChunkAPI();
        MinecraftUserAPI minecraftUserAPI = api.getMinecraftUserAPI();

        /* Get user */
        MinecraftUserDTO minecraftUserDTO = minecraftUserAPI.get(uuid);
        if(minecraftUserDTO == null) throw new Exception("Could not find user when creating land");

        /* Get chunk */
        ChunkDTO chunkDTO = chunkAPI.get(chunk);

        /* If we cannot find one, create new */
        if(chunkDTO == null) {
            Bukkit.getLogger().info("I need a chunk!");
            chunkDTO = chunkAPI.create(chunk, false);
            if(chunkDTO == null) throw new Exception("Could not create chunk when creating land");
        }

        TaggedLocationDTO taggedLocationDTO = new TaggedLocationDTO("spawn", location);

        /* Create the actual land */
        LandDTO landDTO = new LandDTO();
        landDTO.setName(name);
        landDTO.setOwnerId(owner.getUniqueId().toString());
        landDTO.getHomes().add(taggedLocationDTO);

        CompletableFuture<LandDTO> aFuture = fetch.create(landDTO);
        landDTO = aFuture.get();

        if(landDTO == null) throw new Exception("Could not create land");

        DateTime expireDate = new DateTime();
        expireDate.plusMinutes(2);

        String primaryGroup = user.getPrimaryGroup();
        boolean canExpire = primaryGroup.equalsIgnoreCase("default") || primaryGroup.equalsIgnoreCase("barbarian");

        /* Claim the chunk */
        chunkAPI.claim(chunk, landDTO.getId(), canExpire, canExpire ? expireDate : null);

        cachedRecords.put(landDTO.getId(), landDTO);

        return landDTO;
    }

    /**
     * Gets the land via land's id
     * @param id
     * @return LandDTO or null if not found
     */
    public LandDTO get(String id) throws ExecutionException, InterruptedException {
        CompletableFuture<LandDTO> dto = redis.get(id);
        cachedRecords.put(id, dto.get());
        return dto.get();
    }

    /**
     * Gets the land via player's uuid
     * @param player
     * @return LandDTO or null if not found
     */
    public LandDTO get(UUID player) throws ExecutionException, InterruptedException {
        CompletableFuture<LandDTO> dto = redis.get(player.toString());
        cachedRecords.put(player.toString(), dto.get());
        return dto.get();
    }

}

File: src/main/java/com/fendyk/clients/apis/LandAPI.kt
package com.fendyk.clients.apis

import com.fendyk.API
import com.fendyk.DTOs.ChunkDTO
import com.fendyk.DTOs.LandDTO
import com.fendyk.DTOs.TaggedLocationDTO
import com.fendyk.Main
import com.fendyk.clients.ClientAPI
import com.fendyk.clients.fetch.FetchLand
import com.fendyk.clients.redis.RedisLand
import com.fendyk.utilities.Vector2
import net.luckperms.api.model.user.User
import org.bukkit.Chunk
import org.bukkit.Location
import org.bukkit.entity.Player
import org.joda.time.DateTime
import java.util.*
import java.util.concurrent.CompletableFuture

class LandAPI(fetch: FetchLand, redis: RedisLand) : ClientAPI<FetchLand, RedisLand, String, LandDTO>(fetch, redis) {
    companion object {
        var main: Main = Main.instance
        var api: API = main.api
    }

    /**
     * Creates a new land. Requires an existing chunk
     * @param owner
     * @param name
     * @param chunk
     * @return The new land if created or null if something went wrong.
     * @throws Exception When something goes wrong lol
     */
    fun create(owner: Player, name: String, chunk: Chunk, location: Location): CompletableFuture<LandDTO?> {
        return CompletableFuture.supplyAsync {
            val uuid = owner.uniqueId
            val user: User? = Main.instance.luckPermsApi?.userManager?.getUser(uuid)

            // Find chunk
            var chunkDTO: ChunkDTO? = api.chunkAPI.redis.get(Vector2(chunk.x, chunk.z)).join()

            /* If we cannot find one, create new */
            if (chunkDTO == null) {
                chunkDTO = api.chunkAPI.fetch.create(ChunkDTO(chunk.x, chunk.z)).get()
                        ?: throw Exception("Could not create chunk when trying to creating land")
            }

            val taggedLocationDTO = TaggedLocationDTO("spawn", location)

            /* Create the actual land */
            val newLandDTO = LandDTO()
            newLandDTO.name = name
            newLandDTO.ownerId = owner.uniqueId.toString()
            newLandDTO.homes.add(taggedLocationDTO)

            val createdLandDTO: LandDTO = fetch.create(newLandDTO).join() ?: throw Exception("Could not create land")

            val expireDate = DateTime()
            expireDate.plusMinutes(2)
            val primaryGroup = user?.primaryGroup
            val canExpire: Boolean = primaryGroup.equals("default", ignoreCase = true) || primaryGroup.equals("barbarian", ignoreCase = true)

            /* Claim the chunk */
            val isClaimed: Boolean = api.chunkAPI.claim(chunk, newLandDTO.id, canExpire, if (canExpire) expireDate else null).join()
                    ?: throw Exception("Could not claim the chunk you're standing in.")
            cachedRecords[newLandDTO.id] = newLandDTO
            return@supplyAsync newLandDTO
        }
    }

    /**
     * Gets the land via land's id
     * @param id
     * @return LandDTO or null if not found
     */
    fun get(id: String): CompletableFuture<LandDTO?> {
        return CompletableFuture.supplyAsync {
            val landDTO: LandDTO? = redis.get(id).join()
            if (landDTO != null) cachedRecords[id] = landDTO
            return@supplyAsync landDTO
        }
    }

    /**
     * Gets the land via player's uuid
     * @param player
     * @return LandDTO or null if not found
     */
    fun get(player: UUID): CompletableFuture<LandDTO?> {
        return CompletableFuture.supplyAsync {
            val landDTO: LandDTO? = redis.get(player.toString()).join()
            if (landDTO != null) cachedRecords[landDTO.id] = landDTO
            return@supplyAsync landDTO
        }
    }
}